### PR TITLE
feat: datafile accessor

### PR DESCRIPTION
**Pull request by ozayr-zaviar**
## Summary
- toDatafile() method added in DatafileProjectConfig.
- getDatafile() method added in OptimizelyConfig

## Test plan
-  Unit tests added.
- FSC continues to pass.
https://travis-ci.com/github/optimizely/fullstack-sdk-compatibility-suite/builds/181695026

### CODE REVIEW

**Comment by aliabbasrizvi** [comment]
@ozayr-zaviar can you take care of signing our CLA mentioned in the CONTRIBUTING document?

File: src/Optimizely/Config/ProjectConfigInterface.php [hunk 1]
```diff
@@ -159,4 +159,11 @@ public function getFeatureVariableFromKey($featureFlagKey, $variableKey);
      * @return boolean A boolean value that indicates if the experiment is a feature test.
      */
     public function isFeatureExperiment($experimentId);
+
+    /**
+     * Determines if given experiment is a feature test.
```
Comment (oakbani) [inline src/Optimizely/Config/ProjectConfigInterface.php:164]: nit. Fix this docstring

File: src/Optimizely/OptimizelyConfig/OptimizelyConfigService.php [hunk 2]
```diff
@@ -63,6 +68,7 @@ public function __construct(ProjectConfigInterface $projectConfig)
         $this->experiments = $projectConfig->getAllExperiments();
         $this->featureFlags = $projectConfig->getFeatureFlags();
         $this->revision = $projectConfig->getRevision();
+        $this->datafile = $$projectConfig->toDatafile();
```
Comment (oakbani) [inline src/Optimizely/OptimizelyConfig/OptimizelyConfigService.php:71]: nit $$

File: src/Optimizely/OptimizelyConfig/OptimizelyConfig.php [hunk 3]
```diff
@@ -37,11 +37,18 @@ class OptimizelyConfig implements \JsonSerializable
...
     private $featuresMap;
 
-    public function __construct($revision, array $experimentsMap, array $featuresMap)
+    /**
+     * @var string Contents of datafile.
+     */
+    private $datafile;
+    
+
+    public function __construct($revision, array $experimentsMap, array $featuresMap, $datafile = null)
```
Comment (oakbani) [inline src/Optimizely/OptimizelyConfig/OptimizelyConfig.php:46]: do not provide a default null for datafile.

File: src/Optimizely/OptimizelyConfig/OptimizelyConfigService.php (hunk 2 above)
Comment (oakbani) [reply src/Optimizely/OptimizelyConfig/OptimizelyConfigService.php:71]: You haven't passed datafile to new OptimizelyConfig below

File: tests/ConfigTests/DatafileProjectConfigTest.php [hunk 4]
```diff
@@ -14,709 +14,464 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
+namespace Optimizely\Tests;
```
Comment (oakbani) [inline tests/ConfigTests/DatafileProjectConfigTest.php:17]: Please fix this file. This should only show diff of the new unit test you have added.

File: tests/OptimizelyConfigTests/OptimizelyConfigServiceTest.php [hunk 5]
```diff
@@ -467,4 +467,11 @@ public function testJsonEncodeofOptimizelyConfig()
 
         $this->assertEquals(json_encode(json_decode($expectedJSON)), json_encode($response));
     }
+
+    public function testGetDatafile()
+    {
+        $expectedDatafile = DATAFILE_FOR_OPTIMIZELY_CONFIG;
+        $actualDatafile = $this->optConfigService->getDatafile();
```
Comment (oakbani) [inline tests/OptimizelyConfigTests/OptimizelyConfigServiceTest.php:474]: getDatafile() has been defined for OptimizelyConfig, not OptimizelyConfigService.

**Review by oakbani — CHANGES_REQUESTED** [review]
Please address.

**Comment by oakbani** [comment]
@msohailhussain  Testapp PR should be merged to get FSC passed https://github.com/optimizely/php-testapp/pull/65

File: src/Optimizely/Config/DatafileProjectConfig.php [hunk 6]
```diff
@@ -88,6 +88,11 @@ class DatafileProjectConfig implements ProjectConfigInterface
      */
     private $_botFiltering;
 
+    /**
+     * @var string datafile.
+     */
+    private $datafile;
```
Comment (msohailhussain) [inline src/Optimizely/Config/DatafileProjectConfig.php:94]: shouldn't it be $_datafile?

File: src/Optimizely/OptimizelyConfig/OptimizelyConfig.php [hunk 7]
```diff
@@ -37,11 +37,18 @@ class OptimizelyConfig implements \JsonSerializable
      */
     private $featuresMap;
 
-    public function __construct($revision, array $experimentsMap, array $featuresMap)
+    /**
+     * @var string Contents of datafile.
+     */
+    private $datafile;
```
Comment (msohailhussain) [inline src/Optimizely/OptimizelyConfig/OptimizelyConfig.php:43]: same here.

File: src/Optimizely/Config/DatafileProjectConfig.php (hunk 6 above)
Comment (zashraf1985) [inline src/Optimizely/Config/DatafileProjectConfig.php:94]: This should probably be `$_datafile` to be consistent with other private variable conventions. @oakbani ?

File: src/Optimizely/OptimizelyConfig/OptimizelyConfig.php [hunk 8]
```diff
@@ -37,11 +37,18 @@ class OptimizelyConfig implements \JsonSerializable
...
+    
+
+    public function __construct($revision, array $experimentsMap, array $featuresMap, $datafile)
```
Comment (msohailhussain) [inline src/Optimizely/OptimizelyConfig/OptimizelyConfig.php:46]: won't it break existing functionality?

File: tests/OptimizelyConfigTests/OptimizelyEntitiesTest.php [hunk 9]
```diff
@@ -31,7 +31,8 @@ public function testOptimizelyConfigEntity()
         $optConfig = new OptimizelyConfig(
             "20",
             ["a" => "apple"],
-            ["o" => "orange"]
+            ["o" => "orange"],
+            null
```
Comment (msohailhussain) [inline tests/OptimizelyConfigTests/OptimizelyEntitiesTest.php:35]: this shouldn't be changed.

File: tests/OptimizelyConfigTests/OptimizelyEntitiesTest.php [hunk 10]
```diff
@@ -41,7 +42,8 @@ public function testOptimizelyConfigEntity()
         $expectedJson = '{
             "revision": "20",
             "experimentsMap" : {"a": "apple"},
-            "featuresMap": {"o": "orange"}
+            "featuresMap": {"o": "orange"},
+            "datafile": null
```
Comment (msohailhussain) [inline tests/OptimizelyConfigTests/OptimizelyEntitiesTest.php:46]: same here.

File: src/Optimizely/Config/DatafileProjectConfig.php (hunk 6 above)
Comment (aliabbasrizvi) [reply src/Optimizely/Config/DatafileProjectConfig.php:94]: I think the privacy convention has changed. I think @oakbani had showed me that previously. See: https://www.php-fig.org/psr/psr-12/#43-properties-and-constants

File: src/Optimizely/OptimizelyConfig/OptimizelyConfig.php (hunk 7 above)
Comment (aliabbasrizvi) [reply src/Optimizely/OptimizelyConfig/OptimizelyConfig.php:43]: https://www.php-fig.org/psr/psr-12/#43-properties-and-constants

File: src/Optimizely/OptimizelyConfig/OptimizelyConfigService.php [hunk 11]
```diff
@@ -37,6 +37,11 @@ class OptimizelyConfigService
      */
     private $revision;
 
+    /**
+     * @var string config datafile.
```
Comment (aliabbasrizvi) [inline src/Optimizely/OptimizelyConfig/OptimizelyConfigService.php:41]: nit. `@var string String denoting datafile.`

**Pull request merged**